### PR TITLE
Transaction test rewrite

### PR DESCRIPTION
**Pull request by hououou**
- Expand test_framework to support 
  - multiple connection test 
  - set checkpoint_time_out 
  - reload database 
- Rewrite manual transaction set test (done) and int_delete_create_transaction (in process)

### CODE REVIEW

File: test/test_files/transaction/set_transaction.test [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+---- ok
+
+-STATEMENT MATCH (a:person) WHERE a.ID=0 RETURN a.age;
+---- 1
+
+#Timeout
+-CASE TimeoutErrorTest
+-LOG Open Read Only Transaction Triggers Timeout Error For Write Transaction
+-CHECKPOINT_WAIT_TIMEOUT 10000
+-STATEMENT [conn1] BEGIN READ TRANSACTION;
```
Comment (andyfengHKU) [inline test/test_files/transaction/set_transaction.test:40]: I wonder if it's better to create `conn1` first and then use it. Not it seems first looking for `conn1`, if it doesn't exist create one and use it.

File: test/graph_test/graph_test.cpp [hunk 2]
```diff
@@ -157,6 +157,31 @@ void TestHelper::executeScript(const std::string& cypherScript, Connection& conn
         }
     }
 }
+void BaseGraphTest::createDB(uint64_t checkpointWaitTimeout) {
```
Comment (andyfengHKU) [inline test/graph_test/graph_test.cpp:160]: Let's also try to have only 2 functions.

```
createDB(checkpointWaitTimeout)
createConn(checkpointWaitTimeout)
```
 If checkpointWaitTimeout = 0, the behaviour will fall back to createDBAndConn()

File: test/include/graph_test/graph_test.h [hunk 3]
```diff
@@ -162,5 +171,18 @@ class DBTest : public BaseGraphTest {
...
 };
 
+// This class starts multiple database in on-disk mode.
+class MultipleDBTest : public BaseGraphTest {
+public:
+    inline void runTest(std::vector<std::unique_ptr<TestStatement>>& statements, bool flag) {
+        for (auto& statement : statements) {
+            auto conn_name = *statement->conn_name;
+            createConn(conn_name);
+            if (flag)
```
Comment (andyfengHKU) [inline test/include/graph_test/graph_test.h:181]: Is this flag duplicate with the initFlag in `initGraphIfNecessary`?

File: test/runner/e2e_test.cpp [hunk 4]
```diff
@@ -56,6 +58,58 @@ class EndToEndTest : public DBTest {
...
+    std::string getInputDir() override { return dataset + "/"; }
+
+private:
+    TestGroup::DatasetType datasetType;
+    std::string dataset;
+    std::string parquetTempDatasetPath;
+    uint64_t bufferPoolSize;
+    uint64_t checkpointWaitTimeout;
+    std::vector<std::unique_ptr<TestStatement>> testStatements;
+    bool init_graph;
```
Comment (andyfengHKU) [inline test/runner/e2e_test.cpp:103]: Let's be consistent and use camel case, so initGraph

**Review by andyfengHKU — CHANGES_REQUESTED** [review]

File: test/include/graph_test/graph_test.h (hunk 3 above)
Comment (hououou) [reply test/include/graph_test/graph_test.h:181]: this flag respects the dataset is not null, while the initFlag in `initGraphIfNecessary` respects whether the database inits the graph

File: test/runner/e2e_test.cpp (hunk 4 above)
Comment (hououou) [reply test/runner/e2e_test.cpp:103]: ok

File: test/graph_test/graph_test.cpp (hunk 2 above)
Comment (hououou) [reply test/graph_test/graph_test.cpp:160]: ok

File: test/test_files/transaction/set_transaction.test (hunk 1 above)
Comment (hououou) [reply test/test_files/transaction/set_transaction.test:40]: ok

File: test/test_runner/test_parser.cpp [hunk 5]
```diff
@@ -327,5 +338,19 @@ void TestParser::tokenize() {
     }
 }
 
+bool TestParser::extractConnName(std::string& query, TestStatement* statement) {
+    std::regex pattern(R"(\[(conn.*?)\]\s*(.*))");
```
[the remark this answers is not in the document]
Comment (hououou) [reply test/test_runner/test_parser.cpp:342]: I just want to differentiate it from queries such as `"Match (a) where a.listprop=[2,3,4,5,6]"` or `"Match (a)->[r]->(b)"` evolving `"[ ]"`. Do not know how to recognize a connection name better.

File: test/test_runner/test_runner.cpp [hunk 6]
```diff
@@ -21,6 +21,14 @@ void TestRunner::runTest(const std::vector<std::unique_ptr<TestStatement>>& stat
     }
 }
 
+void TestRunner::runTest(
```
[the remark this answers is not in the document]
Comment (hououou) [reply test/test_runner/test_runner.cpp]: has rewritten this function

File: test/include/graph_test/graph_test.h [hunk 7]
```diff
@@ -157,10 +164,24 @@ class DBTest : public BaseGraphTest {
         initGraph();
     }
 
-    inline void runTest(const std::vector<std::unique_ptr<TestStatement>>& statements) {
-        TestRunner::runTest(statements, *conn, databasePath);
+    inline void runTest(const std::vector<std::unique_ptr<TestStatement>>& statements,
```
Comment (andyfengHKU) [inline test/include/graph_test/graph_test.h:167]: put implementation in cpp.

Personally I'm not a big fun of using default parameters. Usually I just overload the function, i.e.
```
runTest(statements);
runTest(statements, checkpointWaitTimeout, connNames);
```
But you can decide this.

File: test/include/graph_test/graph_test.h [hunk 8]
```diff
@@ -157,10 +164,24 @@ class DBTest : public BaseGraphTest {
...
+        for (auto& statement : statements) {
+            if (statement->reLoadDBFlag) {
+                createDB(checkpointWaitTimeout);
+                createConns(connNames);
+                continue;
+            }
+            if (conn) {
+                TestRunner::runTest(statement.get(), *conn, databasePath);
+            } else {
+                auto conn_name = *statement->conn_name;
```
Comment (andyfengHKU) [inline test/include/graph_test/graph_test.h:180]: use camel case. connName

File: test/include/test_runner/test_group.h [hunk 9]
```diff
@@ -18,6 +20,9 @@ struct TestStatement {
     bool enumerate = false;
     bool checkOutputOrder = false;
     std::string expectedTuplesCSVFile;
+    // for multiple conns
+    std::optional<std::string> conn_name; // default conntion name conn0
```
Comment (andyfengHKU) [inline test/include/test_runner/test_group.h:24]: ditto. camel case. And I believe default name is "default"? Replace default with a constant and apply across this PR

File: test/include/test_runner/test_group.h [hunk 10]
```diff
@@ -18,6 +20,9 @@ struct TestStatement {
...
+    bool reLoadDBFlag = false;
```
Comment (andyfengHKU) [inline test/include/test_runner/test_group.h:25]: reLoadDBFlag -> reloadDBFlag

File: test/include/test_runner/test_group.h [hunk 11]
```diff
@@ -31,6 +36,12 @@ struct TestGroup {
     std::unordered_map<std::string, std::vector<std::unique_ptr<TestStatement>>>
         testCasesStatementBlocks;
     uint64_t bufferPoolSize = common::BufferPoolConstants::DEFAULT_BUFFER_POOL_SIZE_FOR_TESTING;
+    // for multiple connections
+    uint64_t checkpointWaitTimeout =
+        common::DEFAULT_CHECKPOINT_WAIT_TIMEOUT_FOR_TRANSACTIONS_TO_LEAVE_IN_MICROS;
+    ;
```
Comment (andyfengHKU) [inline test/include/test_runner/test_group.h:42]: remove line

File: test/include/test_runner/test_group.h [hunk 12]
```diff
@@ -31,6 +36,12 @@ struct TestGroup {
     std::unordered_map<std::string, std::vector<std::unique_ptr<TestStatement>>>
         testCasesStatementBlocks;
     uint64_t bufferPoolSize = common::BufferPoolConstants::DEFAULT_BUFFER_POOL_SIZE_FOR_TESTING;
+    // for multiple connections
+    uint64_t checkpointWaitTimeout =
```
Comment (andyfengHKU) [inline test/include/test_runner/test_group.h:40]: This is not really for multiple connections so move this line above the comment. Also @ray6080 should comment if this config will exist in the long term. I feel we might remove it at some point.

File: test/test_files/transaction/set_transaction.test [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+Alice
+
+#VeryLongStringErrorsTest
+-CASE VeryLongStringErrorsTest
+-LOG Set Very Long String Errors Test
+-STATEMENT BEGIN WRITE TRANSACTION;
+---- ok
+-DEFINE test_long_string REPEAT 4096 "a"
+-STATEMENT MATCH (a:person) WHERE a.ID=0 SET a.fName='${test_long_string}';
+---- ok
```
Comment (andyfengHKU) [inline test/test_files/transaction/set_transaction.test:79]: new line

**Review by andyfengHKU — APPROVED** [review]
Also remove the cpp test files after migrating

File: test/include/test_runner/test_group.h (hunk 9 above)
Comment (ray6080) [reply test/include/test_runner/test_group.h:24]: @hououou did you address this comment?

File: test/include/test_runner/test_group.h (hunk 10 above)
Comment (ray6080) [reply test/include/test_runner/test_group.h:25]: @hououou Ditto. I don't see this comment addressed.

File: test/include/test_runner/test_group.h (hunk 12 above)
Comment (ray6080) [reply test/include/test_runner/test_group.h:40]: I think this will exist for quite a long time.

**Pull request merged**